[Console] Add domain exceptions to replace generic exceptions

GromNaN opened the pull request:
Creates domain specific exception classes for the case where a user type an invalid command name or option name.

| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #14873 |
| License | MIT |
| Doc PR | N/A |

TODO:
- [x] Replace `\InvalidArgumentException` by `Symfony\Component\Console\Exception\InvalidArgumentException`
- [x] Add `Symfony\Component\Console\Exception\ExceptionInterface`

GromNaN committed:
[Console] Add CommandNotDefinedException to replace generic \InvalidArgumentException

GromNaN committed:
[Console] Add InvalidOptionException to replace generic \InvalidArgumentException

GromNaN commented on src/Symfony/Component/Console/Tests/Fixtures/application_renderexception4.txt:
```diff
@@ -1,9 +1,9 @@
+                                                                    
+  [Symfony\Component\Console\Exception\CommandNotDefinedException]  
+  Command "foo" is not define                                       
+  d.                                                                
```
I'm not sure this is the correct behavior. The line is wrapped, but not the class name, and the rest of the line is filled with spaces.

fabpot replied:
I think `d` should be on the same line as `define`.

GromNaN commented:
Any idea how to fix this failing test ?
https://travis-ci.org/symfony/symfony/jobs/65716176#L438

phansys commented:
AFAIK, the alternative names was originally [ordered](https://github.com/symfony/symfony/blob/99ec717a22b811f2a46a82b3c49889c87cd76db2/src/Symfony/Component/Console/Application.php#L1090) with ```levenshtein()```, but I don't know if your issue is related to this.

phansys commented on src/Symfony/Component/Console/Exception/CommandNotDefinedException.php:
```diff
@@ -0,0 +1,41 @@
+    {
+        return $this->alternatives;
+    }
+}
```
Please, check missing new lines at EOFs.

GromNaN replied:
Fixed.

GromNaN committed:
Use CommandNotDefinedException in ApplicationDescription

fabpot commented on src/Symfony/Component/Console/Tests/ApplicationTest.php:
```diff
@@ -176,7 +176,7 @@ public function testSilentHelp()
 
     /**
-     * @expectedException        \InvalidArgumentException
+     * @expectedException        Symfony\Component\Console\Exception\CommandNotDefinedException
```
Missing leading \ on all `@expectedException` tags

fabpot replied:
Can you fix this? We are using a leading `\` everywhere else in Symfony for such PHPUnit annotations.

fabpot commented on src/Symfony/Component/Console/Exception/CommandNotDefinedException.php:
```diff
@@ -0,0 +1,41 @@
+/**
+ * Represents an incorrect command name typed in the console.
+ */
+class CommandNotDefinedException extends \InvalidArgumentException
```
You should also create a marker like for most other components like the one in the routing component `Symfony\Component\Routing\Exception\ExceptionInterface` and then implement it here:

`class CommandNotDefinedException extends \InvalidArgumentException implements ExceptionInterface`

That way, it's possible to catch all exceptions coming from the Console.

fabpot commented:
As you are moving exceptions to specific ones, what about doing it for the whole component?

fabpot commented:
@GromNaN Any news on this one? Can you address the comments?

GromNaN commented:
OK for moving remaining exception to specific ones. Is it necessary to keep the distinction between `InvalidArgumentException` and `LogicException` in classes like [Console/Question/Question](https://github.com/GromNaN/symfony/blob/console-exception/src/Symfony/Component/Console/Question/Question.php#L136) ?

fabpot commented:
Yes, I think we need to keep the distinction here (at least to keep BC).

GromNaN commented:
I'll create `Symfony\Component\Console\Exception\InvalidArgumentException` and  `Symfony\Component\Console\Exception\LogicException`

GromNaN committed:
[Console] Use domain specific exceptions

Replace exception thrown by specific exception implementing the
same interface.

GromNaN commented:
Rebased and made all changes.

I still have this error :
```
1) Symfony\Component\Console\Tests\ApplicationTest::testFindAlternativeNamespace
Failed asserting that Array &0 (
    0 => 'foo'
    1 => 'foo1'
    2 => 'foo3'
) is identical to Array &0 (
    0 => 'foo3'
    1 => 'foo1'
    2 => 'foo'
).
```

fabpot commented on src/Symfony/Component/Console/Exception/CommandNotDefinedException.php:
```diff
@@ -0,0 +1,43 @@
+ *
+ * @author Jérôme Tamarelle <jerome@tamarelle.net>
+ */
+class CommandNotDefinedException extends \InvalidArgumentException implements ExceptionInterface
```
I would have named it `CommandNotFoundException` to be more consistent with other exception names in other components.

GromNaN replied:
That's close to the `RouteNotFoundException` in the Routing component. Have you something in mind ?

fabpot replied:
`CommandNotFoundException`

fabpot commented on src/Symfony/Component/Console/Helper/QuestionHelper.php:
```diff
@@ -11,13 +11,14 @@
 use Symfony\Component\Console\Formatter\OutputFormatterStyle;
 use Symfony\Component\Console\Question\Question;
 use Symfony\Component\Console\Question\ChoiceQuestion;
-
```
should be reverted

fabpot commented:
I made some minor comments, but I'm :+1: on this change.

ping @symfony/deciders

GromNaN commented on src/Symfony/Component/Console/Helper/SymfonyQuestionHelper.php:
```diff
@@ -38,7 +39,7 @@ public function ask(InputInterface $input, OutputInterface $output, Question $qu
             // make required
             if (!is_array($value) && !is_bool($value) && 0 === strlen($value)) {
-                throw new \Exception('A value is required.');
+                throw new LogicException('A value is required.');
```
Here I've replaced the only instance of basic ` \Exception` by a more precise `LogicException`.

xabbuh commented on src/Symfony/Component/Console/Exception/ExceptionInterface.php:
```diff
@@ -0,0 +1,23 @@
+ *
+ * @api
+ */
+interface ExceptionInterface
```
Do we really need the interface? And do we need to tag it with as an `@api` interface given that the exception class are not tagged this way?

fabpot replied:
It needs to be consistent with the way we manage exception in the other components. We can indeed remove the `@api` tag but I'm not sure this will ever change.

GromNaN committed:
Rename CommandNotDefinedException to CommandNotFoundException for consistency

dunglas commented:
:+1:

fabpot commented:
Thank you @GromNaN.

xabbuh commented:
@GromNaN I had a look at the failing tests: #15940
